### ISSUE TITLE
Find interface automatically error

### ISSUE THREAD

**Issue by donpedro27476** [sophron/wifiphisher#213]
After running the script, everything works normally. However, immediately after getting the message- "fake access point initializing", error message appears- " Driver initialization failed! " (hostapd error). Then it says "try a different wireless interface using -aI option" . 

Information:
Device- Nexus 4
Wireless card- tplink tl-wn722n (using USB y cable and power back)

**Comment by blackHatMonkey** [comment]
For ```wifiphisher``` to work you need to have at least one card with ```AP``` mode available. Please check which of you wireless adapters supports ```AP``` mode and then specify it when running ```wifiphisher```.
For instance if ```wlan0``` supports ```AP``` mode then run the fallowing:
```bash
sudo python wifiphisher -aI wlan0
```
Let me know if you have any problems.

**Comment by donpedro27476** [comment]
Thanks for the help @blackhatMonkey
It works now.

**Comment by donpedro27476** [comment]
![screenshot_2016-02-23-23-24-22](https://cloud.githubusercontent.com/assets/17426160/13261442/15b659ba-da86-11e5-9769-fcd1efb59361.png)

**Comment by donpedro27476** [comment]
So it works to the extent that the Fake AP is created. The victim device however shows " error-name-not-resolved" when the 'sign in to wifi xx' is clicked. 

It also appears that the second wifi seems to go offline wherever the Fake AP is running

**Comment by donpedro27476** [comment]
So it works to the extent that the Fake AP is created. The victim device however shows " error-name-not-resolved" when the 'sign in to wifi xx' is clicked. 

It also appears that the second wifi seems to go offline wherever the Fake AP is running

**Comment by blackHatMonkey** [comment]
To be honest ```wifiphisher``` is developed to be run on ```linux kali``` and since I don't have the hardware to test, I'm afraid I can't help you. However something you may want to try is to do a hard reset on the android device. Also you may want to try connecting from a different device and check if the same error is presented.

**Comment by crucio95** [comment]
using a usb wifi adapter Alfa AWUS036NHA that supports AP mode, and i've tried it with airmon and it works, i get the drive initialization error. can you help me??

**Comment by blackHatMonkey** [comment]
Can you post the contents of ```/tmp/wifiphisher-hostapd.log``` while running the script?

**Comment by donpedro27476** [comment]
Configuration file: /tmp/hostapd.conf
Using interface wlan1 with hwaddr 60:e6:27:12:fb:d3 and ssid "xyz"wlan1: interface state UNINITIALIZED->ENABLED
wlan1: AP-ENABLED
rfkill: WLAN unblocked

**Comment by donpedro27476** [comment]
The above is the log using the tplink tl-wn 722n

**Comment by blackHatMonkey** [comment]
Can you try using the latest version of ```wifiphisher```?

**Comment by donpedro27476** [comment]
![screenshot_2016-04-17-12-30-10](https://cloud.githubusercontent.com/assets/17426160/14585602/42906bec-0498-11e6-9938-e9cf2370bb58.png)

**Comment by donpedro27476** [comment]
I downloaded the latest version of wifiphisher and tried running the script using both the original command you'd suggested 
sudo python wifiphisher.py -aI wlan1
(Results in the message that there is no file named as wifiphisher.py in the directory)

As well as python bin/wifiphisher 

As you will notice above I even tried using sudo and the -aI option. Both result in the same message as displayed above. I still use the TP-Link tl-wn722n. Notice above I tried running wifite and it works perfectly.

**Comment by blackHatMonkey** [comment]
Can you try running 
```bash
sudo python bin/wifiphisher
```
and see if you get the same error.

**Comment by sophron** [comment]
Hi @donpedro27476,

Wifiphisher has not been tested on nethunter from our side so we are expecting issues there.

Can you please hit `iw dev` in the terminal? Then for each physical interface, try `iw <physical_interface> info` and paste here the output.  That will helps us trace the problem.

**Comment by donpedro27476** [comment]
Hi @blackHatMonkey
As requested please see the following.....

root@kali:~# sudo python bin/wifiphisher
python: can't open file 'bin/wifiphisher': [Errno 2] No such file or directory

The following error appears after using sudo python bin/wifiphisher


@sophron as suggested please find the output requested


root@kali:~# iw dev
phy#1
        Interface wlan1
                ifindex 28
                type managed
phy#0
        Interface p2p0
                ifindex 27
                type managed
        Interface wlan0
                ifindex 26
                type managed


root@kali:~# iw wlan1 info
Interface wlan1
        ifindex 28
        type managed
        wiphy 1


root@kali:~# iw phy#1 info
Wiphy phy1
        Band 1:
                Capabilities: 0x116e
                        HT20/HT40
                        SM Power Save disabled
                        RX HT20 SGI
                        RX HT40 SGI
                        RX STBC 1-stream
                        Max AMSDU length: 3839 bytes
                        DSSS/CCK HT40
                Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
                Minimum RX AMPDU time spacing: 8 usec (0x06)
                HT TX/RX MCS rate indexes supported: 0-7
                Frequencies:
                        * 2412 MHz [1] (20.0 dBm)
                        * 2417 MHz [2] (20.0 dBm)
                        * 2422 MHz [3] (20.0 dBm)
                        * 2427 MHz [4] (20.0 dBm)
                        * 2432 MHz [5] (20.0 dBm)
                        * 2437 MHz [6] (20.0 dBm)
                        * 2442 MHz [7] (20.0 dBm)
                        * 2447 MHz [8] (20.0 dBm)
                        * 2452 MHz [9] (20.0 dBm)
                        * 2457 MHz [10] (20.0 dBm)
                        * 2462 MHz [11] (20.0 dBm)
                        * 2467 MHz [12] (20.0 dBm)
                        * 2472 MHz [13] (20.0 dBm)
                        * 2484 MHz [14] (disabled)
                Bitrates (non-HT):
                        * 1.0 Mbps
                        * 2.0 Mbps (short preamble supported)
                        * 5.5 Mbps (short preamble supported)
                        * 11.0 Mbps (short preamble supported)
                        * 6.0 Mbps
                        * 9.0 Mbps
                        * 12.0 Mbps
                        * 18.0 Mbps
                        * 24.0 Mbps
                        * 36.0 Mbps
                        * 48.0 Mbps
                        * 54.0 Mbps
        max # scan SSIDs: 4
        max scan IEs length: 2257 bytes
        Retry short limit: 7
        Retry long limit: 4
        Coverage class: 0 (up to 0m)
        Supported Ciphers:
                * WEP40 (00-0f-ac:1)
                * WEP104 (00-0f-ac:5)
                * TKIP (00-0f-ac:2)
                * CCMP (00-0f-ac:4)
        Available Antennas: TX 0 RX 0
        Supported interface modes:
                 * IBSS
                 * managed
                 * AP
                 * AP/VLAN
                 * monitor
                 * P2P-client
                 * P2P-GO
        software interface modes (can always be added):
                 * AP/VLAN
                 * monitor
        interface combinations are not supported
        Supported commands:
                 * new_interface
                 * set_interface
                 * new_key
                 * start_ap
                 * new_station
                 * set_bss
                 * authenticate
                 * associate
                 * deauthenticate
                 * disassociate
                 * join_ibss
                 * remain_on_channel
                 * set_tx_bitrate_mask
                 * frame
                 * frame_wait_cancel
                 * set_wiphy_netns
                 * set_channel
                 * set_wds_peer
                 * probe_client
                 * set_noack_map
                 * register_beacons
                 * connect
                 * disconnect
        Supported TX frame types:
                 * IBSS: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * managed: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * AP: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * AP/VLAN: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * mesh point: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * P2P-client: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * P2P-GO: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
        Supported RX frame types:
                 * IBSS: 0xd0
                 * managed: 0x40 0xd0
                 * AP: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
                 * AP/VLAN: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
                 * mesh point: 0xb0 0xc0 0xd0
                 * P2P-client: 0x40 0xd0
                 * P2P-GO: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
        Device supports RSN-IBSS.
        HT Capability overrides:
                 * MCS: ff ff ff ff ff ff ff ff ff ff
                 * maximum A-MSDU length
                 * supported channel width
                 * short GI for 40 MHz
                 * max A-MPDU length exponent
                 * min MPDU start spacing
        Device supports TX status socket option.
        Device supports HT-IBSS.
root@kali:~#

**Comment by blackHatMonkey** [comment]
@donpedro27476 Thanks for the info. I will try to reproduce it and let you know what happens.

**Comment by sophron** [comment]
@blackHatMonkey Two notes:

1) This <a href="https://github.com/sophron/wifiphisher/blob/master/wifiphisher/interfaces.py#L447-L451">block here</a> should be rewritten with regex in mind. Furthermore, you should not match the exact line but instead you should check if the line contains specific values, such as "AP" or "monitor".

2) The if statement in <a href="https://github.com/sophron/wifiphisher/blob/master/wifiphisher/interfaces.py#L610">L610</a> is wrong. What happens if you have only 1 interface that supports both AP and monitor modes? It will be removed from both `ap_available` and `monitor_available` (so both lists are now empty) but then the script will try to retrieve an interface from an empty list on L620.

**Comment by blackHatMonkey** [comment]
@sophron 
1) If i recall correctly the reason I did that was because there are some other lines in the ```iw info``` return that would contain ```AP``` even though the interface doesn't support it. I'll have to double check but in that case using regex would be difficult.

2) Yes there is case that I hadn't considered when I wrote it and that is if we have two interfaces where one supports both ```AP``` and ```monitor``` mode and the other supports neither. In that case the error will happen. However that shouldn't be the case here since the OP has two ```AP``` interfaces and one ```monitor``` so it should have executed <a href="https://github.com/sophron/wifiphisher/blob/master/wifiphisher/interfaces.py#L610">L622</a>.
